File: cab/cab/Models/Booking.swift
import Foundation

// MARK: - Flexible ID-or-Object decoding

/// Handles API fields that can be returned as either a plain String ID or a full populated object.
enum IDOrObject<T: Codable>: Codable {
    case id(String)
    case object(T)

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let object = try? container.decode(T.self) {
            self = .object(object)
        } else if let id = try? container.decode(String.self) {
            self = .id(id)
        } else {
            throw DecodingError.typeMismatch(
                IDOrObject.self,
                .init(codingPath: decoder.codingPath, debugDescription: "Expected String or object")
            )
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .id(let id): try container.encode(id)
        case .object(let obj): try container.encode(obj)
        }
    }

    var object: T? {
        if case .object(let obj) = self { return obj }
        return nil
    }
}

// MARK: - BookingStatus

enum BookingStatus: String, CaseIterable {
    case pending   = "pending"
    case confirmed = "confirmed"
    case completed = "completed"
    case cancelled = "cancelled"

    var displayName: String {
        rawValue.capitalized
    }
}

// MARK: - Booking

struct Booking: Codable, Identifiable {
    let id: String
    let customerId: IDOrObject<User>
    let routeId: IDOrObject<Route>
    let travelDate: String
    let numberOfPeople: Int
    let preferredSeater: Int
    let prefersCNG: Bool
    let status: String
    let assignedCabId: IDOrObject<Cab>?
    let totalAmount: Double
    let customerNotes: String?
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case customerId, routeId, travelDate, numberOfPeople, preferredSeater, prefersCNG
        case status, assignedCabId, totalAmount, customerNotes, createdAt
    }

    var route: Route? { routeId.object }
    var assignedCab: Cab? { assignedCabId?.object }
    var customer: User? { customerId.object }

    var statusEnum: BookingStatus { BookingStatus(rawValue: status) ?? .pending }
}

// MARK: - Request body

struct CreateBookingRequest: Encodable {
    let routeId: String
    let travelDate: String
    let numberOfPeople: Int
    let preferredSeater: Int
    let prefersCNG: Bool
    let customerNotes: String?
}

File: backend/src/models/Booking.js
const mongoose = require('mongoose');

const bookingSchema = new mongoose.Schema(
  {
    customerId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    routeId: { type: mongoose.Schema.Types.ObjectId, ref: 'Route', required: true },
    travelDate: { type: Date, required: true },
    numberOfPeople: { type: Number, required: true, min: 1 },
    preferredSeater: { type: Number, required: true, enum: [4, 6, 7] },
    prefersCNG: { type: Boolean, default: false },
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'completed', 'cancelled'],
      default: 'pending',
    },
    assignedCabId: { type: mongoose.Schema.Types.ObjectId, ref: 'Cab', default: null },
    totalAmount: { type: Number, required: true, min: 0 },
    customerNotes: { type: String, trim: true, default: '' },
  },
  { timestamps: true }
);

module.exports = mongoose.model('Booking', bookingSchema);
